### PR TITLE
Check remaining transaction value & make value balance signs match the spec 

### LINKED ISSUE #2381: ZIP-209: Validate transaction value balances - Implementation
#### Transparent

> Transparent inputs to a transaction insert value into a transparent transaction value pool associated with the transaction, and transparent outputs remove value from this pool. As in Bitcoin, the remaining value in the pool is available to miners as a fee.
>
> Consensus rule: The remaining value in the transparent transaction value pool MUST be nonnegative.

https://zips.z.cash/protocol/protocol.pdf#transactions

#### Sprout

> vold takes value from the transparent transaction value pool and vnew adds value to the transparent transaction
value pool . As a result, vold is treated like an output value, whereas vnew is treated like an input value.

> As stated in § 4.3 ‘JoinSplit Descriptions’ on p. 36, either vold or vnew MUST be zero.

https://zips.z.cash/protocol/protocol.pdf#joinsplitbalance

#### Sapling

> A positive Sapling balancing value takes value from the Sapling transaction value pool and adds it to the transparent transaction value pool . A negative Sapling balancing value does the reverse. As a result, positive vbalanceSapling is treated like an input to the transparent transaction value pool , whereas negative vbalanceSapling is treated like an output from that pool.

https://zips.z.cash/protocol/protocol.pdf#saplingbalance

#### Orchard

> vbalanceOrchard is encoded in a transaction as the field valueBalanceOrchard. If a transaction has no Action descriptions, vbalanceOrchard is implicitly zero. Transaction fields are described in § 7.1 ‘Transaction Encoding and Consensus’ on p. 116.
>
> A positive Orchard balancing value takes value from the Orchard transaction value pool and adds it to the transparent transaction value pool . A negative Orchard balancing value does the reverse. As a result, positive vbalanceOrchard is treated like an input to the transparent transaction value pool , whereas negative vbalanceOrchard is treated like an output from that pool.

https://zips.z.cash/protocol/protocol.pdf#orchardbalance

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to check the remaining value in each non-coinbase transaction, to make sure no new coins get created.

We also want to update the value balance signs to match the spec.

We also need to test the transaction and chain value balance checks. (Tickets #2381 and #1895.)
But currently, Zebra generates random values, which are almost always invalid.

### Consensus Rules

- Check remaining transaction value
- Update docs for coinbase remaining value consensus rule, based on recent spec changes

### Implement Design Changes

- Make the value balance signs match the spec (see the design changes in PR #2576)
- Use a NonNegative constraint for transparent values
  - using this constraint correctly represents the value range, and avoids redundant checks for negative values

### Test Changes

- Allocate total transaction input value to outputs in tests
- Add more test methods for modifying transparent values and shielded value balances
- Consistently use ValueBalanceError for ValueBalances & make amount::Errors into specific pool ValueBalanceErrors
  - this allows us to tag value balance errors with the specific pool that caused the error

Closes #2381.

## Review

@oxarbitrage can review this PR, we're working on the value balance implementation together.

This PR might need to be rebased on `main` before it merges, it is currently based on PR #2577.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

We need to check that each chain value pool is non-negative, and generate test data that passes that check.

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2566)

### CODE REVIEW

File: zebra-chain/src/transaction.rs [hunk 1]
```diff
@@ -725,11 +706,163 @@ impl Transaction {
...
+                ..
+            }
+            | Transaction::V5 { .. } => Box::new(std::iter::empty()),
+        }
+    }
+
+    /// Modify the `vpub_new` fields from `JoinSplit`s in this transaction,
+    /// regardless of version.
+    ///
+    /// This value is added to the transparent value pool of this transaction,
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// This value is added to the value pool of this transaction,
```

File: zebra-chain/src/transaction.rs [hunk 2]
```diff
@@ -703,6 +679,11 @@ impl Transaction {
     /// The sum of the outputs spent by transparent inputs in `tx_in` fields,
     /// minus the sum of newly created outputs in `tx_out` fields.
     ///
+    /// Positive values are added to the transparent value pool,
+    /// and removed from the value pool of this transaction.
+    /// Negative values are removed from transparent,
+    /// and removed from this transaction.
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// The sum of newly created outputs in `tx_out` fields,
    /// minus the sum of the outputs spent by transparent inputs in `tx_in` fields.
    ///
    /// Positive values are added to the transparent value pool,
    /// and removed from the value pool of this transaction.
    /// Negative values are removed from transparent,
    /// and added to this transaction.
```

File: zebra-chain/src/transaction.rs [hunk 3]
```diff
@@ -725,11 +706,163 @@ impl Transaction {
...
+            }
+            | Transaction::V5 { .. } => Box::new(std::iter::empty()),
+        }
+    }
+
+    /// Modify the `vpub_old` fields from `JoinSplit`s in this transaction,
+    /// regardless of version.
+    ///
+    /// This value is removed from the value pool of this transaction,
+    /// and added to the sprout value pool.
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// See `sprout_pool_added_values` for details.
```

File: zebra-chain/src/transaction.rs [hunk 4]
```diff
@@ -725,11 +706,163 @@ impl Transaction {
...
+
     /// Return the sprout value balance
     ///
     /// The change in the sprout value pool.
     /// The sum of all sprout `vpub_old` fields, minus the sum of all `vpub_new` fields.
     ///
+    /// Positive values are added to the sprout value pool,
+    /// and removed from the value pool of this transaction.
+    /// Negative values are removed from sprout,
+    /// and removed from this transaction.
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// and added to this transaction.
```

File: zebra-chain/src/transaction.rs [hunk 5]
```diff
@@ -755,6 +888,11 @@ impl Transaction {
     /// The change in the sapling value pool.
     /// The negation of the sum of all `valueBalanceSapling` fields.
     ///
+    /// Positive values are added to the sapling value pool,
+    /// and removed from the value pool of this transaction.
+    /// Negative values are removed from sapling,
+    /// and removed from this transaction.
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// and added to this transaction.
```

File: zebra-chain/src/transaction.rs [hunk 6]
```diff
@@ -781,11 +919,45 @@ impl Transaction {
...
+
     /// Return the orchard value balance.
     ///
     /// The change in the orchard value pool.
     /// The negation of the sum of all `valueBalanceOrchard` fields.
     ///
+    /// Positive values are added to the orchard value pool,
+    /// and removed from the value pool of this transaction.
+    /// Negative values are removed from orchard,
+    /// and removed from this transaction.
```
Comment (teor2345) [inline zebra-chain/src/transaction.rs]:
```suggestion
    /// and added to this transaction.
```

File: zebra-chain/src/transaction/joinsplit.rs [hunk 7]
```diff
@@ -53,6 +53,13 @@ impl<P: ZkSnarkProof> JoinSplitData<P> {
         std::iter::once(&self.first).chain(self.rest.iter())
     }
 
+    /// Modify the [`JoinSplit`]s in `self`, in the order they appear
+    /// in the transaction.
```
Comment (teor2345) [inline zebra-chain/src/transaction/joinsplit.rs]:
```suggestion
    /// Modify the [`JoinSplit`]s in `self`,
    /// in the order they appear in the transaction.
```

File: zebra-state/src/error.rs [hunk 8]
```diff
@@ -140,11 +140,28 @@ pub enum ValidateContextError {
...
-    InvalidRemainingTransparentValue {
+    InvalidRemainingTransactionValue {
+        amount_error: amount::Error,
+        height: block::Height,
+        tx_index_in_block: usize,
+        transaction_hash: zebra_chain::transaction::Hash,
+    },
+
+    #[error(
+        "error while calculating the remaining value in the transparent transaction value pool: \
```
Comment (teor2345) [inline zebra-state/src/error.rs]:
```suggestion
        "error while calculating the remaining value in the transaction value pool: \
```

File: zebra-state/src/service/check/utxo.rs [hunk 9]
```diff
@@ -86,6 +86,8 @@ pub fn transparent_spend(
         }
     }
 
+    remaining_transaction_value(prepared, &block_spends)?;
+
```
Comment (teor2345) [inline zebra-state/src/service/check/utxo.rs:90]:
```suggestion
```

**Review by teor2345 — COMMENTED** [review]
I want to fix some comments and remove a temporary check

**Comment by teor2345** [comment]
This seems to work, but I need to fix some confusing parts of the code, and maybe split it into a few separate PRs.

**Comment by oxarbitrage** [comment]
> This seems to work, but I need to fix some confusing parts of the code, and maybe split it into a few separate PRs.

Yes, some split here will be valuable.

**Comment by teor2345** [comment]
> > This seems to work, but I need to fix some confusing parts of the code, and maybe split it into a few separate PRs.
> 
> Yes, some split here will be valuable.

I made the signs match the spec, and split some more changes into other PRs.

I also made some other changes to the types and errors so it is easier to write code and diagnose issues.

**Review by oxarbitrage — APPROVED** [review]
> Reviewed 3 of 12 files at r2, 3 of 8 files at r3, 3 of 8 files at r4, 11 of 12 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2566)* status: all files reviewed, 9 unresolved discussions (waiting on @teor2345)

**Pull request merged**